### ISSUE TITLE
Cannot enter/create PSSession from linux to windows machine

### ISSUE THREAD

**Issue by KaloferovLab** [PowerShell/PowerShell#6647]
Steps to reproduce
------------------
From LInux to WIn :
```powershell
enter-PSSession -ConfigurationName powershell.6.1.0-preview.1
```

Expected behavior
-----------------
Enter the pssession on the Windows box. Same error when i try to create new pesssession on the windows box. 
```none

```

Actual behavior
---------------
Error
```none
New-PSSession : MI_RESULT_ACCESS_DENIED
At line:1 char:1
Enter-PSSession -ComputerName <IP> -Credential <username>
  + CategoryInfo          : InvalidOperation: (:) [New-PSSession], PSInvalidOperationException
  + FullyQualifiedErrorId : InvalidOperation,Microsoft.PowerShell.Commands.NewPSSessionCommand
```

Environment data
----------------
WIndows server:
```powershell
Name                           Value
----                           -----
PSVersion                      6.1.0-preview.1
PSEdition                      Core
GitCommitId                    v6.1.0-preview.1
OS                             Microsoft Windows 10.0.14393
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```
LInux Server:
```powershell
Name                           Value
----                           -----
PSVersion                      6.1.0-preview.1
PSEdition                      Core
GitCommitId                    v6.1.0-preview.1
OS                             Linux 3.10.0-514.e17.x86_64 #1 SMP Tue Nov 22 16:42:41 UTC 2016
Platform                       Unix
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

The alpha vesion of the plugin didn't support PSCredential Object and  WinRM. Does it support them now? 
Found this from last year. 
https://github.com/PowerShell/PowerShell/issues/5742
Does cross platform remoting work now in 6.1.0?

**Comment by SteveL-MSFT** [comment]
@KaloferovLab remoting over WSMan is supported, but much more limited than what you get with Windows and WinRM.  From Linux, you should use `-Credential` as unlike Windows you can't use the current security context.  Also, use `-Authentication Basic`.

**Comment by brunobml** [comment]
I have tried many times establishing a new session from linux RHEL7 to windows server 2016.
Always get the same error.
I also tried , athentication Kerberos, Basic, Negotiable, etc....

Enter-PSSession : MI_RESULT_ACCESS_DENIED
At line:1 char:1
+ Enter-PSSession -ComputerName WSRVPRD001 -Credential (Get-Credential) ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : InvalidArgument: (WSRVPRD001:String) [Enter-PSSession], PSInvalidOperationException
+ FullyQualifiedErrorId : CreateRemoteRunspaceFailed

**Comment by tekniko24** [comment]
New-PSSession -ComputerName <host> -Credential $credential -Authentication Negotiate

Use Negotiate for authentication. I don't recommend using basic as that would send your password in plain text over the network and you would also need to manually allow the remote host to accept basic authentication as it is not allowed by default. Specifying Negotiate results in using NTLM authentication which can be verified by checking $PSSenderInfo.UserInfo.Identity.AuthenticationType

**Comment by danports** [comment]
If you are seeing this error on Debian or Ubuntu, see  https://github.com/PowerShell/PowerShell/issues/7342#issuecomment-477711182.

**Comment by mgseelan** [comment]
I also facing same problem when I am accessing from ubuntu 16.04, with powershell version and details are included

Name                           Value
----                           -----
PSVersion                      6.2.1
PSEdition                      Core
GitCommitId                    6.2.1
OS                             Linux 4.15.0-1036-gcp #38~16.04.1-Ubuntu SMP Tue Jun 25 15:30:46 UTC 2019
Platform                       Unix
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0


My error output :
PowerShell credential request
Enter your credentials.
Password for user XXXXXXXXXXXX: **********

enter-pssession : MI_RESULT_ACCESS_DENIED
At line:1 char:1
+ enter-pssession <ip> -credential XXXXXXXXXXXX
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : InvalidArgument: (<ip>:String) [Enter-PSSession], PSInvalidOperationException
+ FullyQualifiedErrorId : CreateRemoteRunspaceFailed

**Comment by dheitsc** [comment]
I also face this Issue on Centos 7 and Server 2012r2 / 2016

**Comment by jorioux** [comment]
Same issue on Centos 7 and WinServer 2019

`Invoke-Command` gives the same error.

**Comment by aric49** [comment]
So it appears this issue has been open for a while.   I recently ran into this issue via #10764.  Any plans to get this resolved in future releases?   This holding up some progress to port some Windows automation over to Linux based environment.

**Comment by arnydo** [comment]
Experiencing this ongoing issue as well.
![image](https://user-images.githubusercontent.com/11653079/67111832-d8c0fb80-f1a3-11e9-9707-712a77623b66.png)

# Linux > Windows 2012 R2
```
Enter-PSSession -Credential $creds -ConfigurationName microsoft.exchange -ConnectionUri http://x.x.x.x/powershell -Authentication Negotiate -Verbose
```
```
Enter-PSSession : Connecting to remote server x.x.x.x failed with the following error message : MI_RESULT_ACCESS_DENIED For more information, see the about_Remote_Troubleshooting Help topic.     
```

#  Linux to Exchange 2013
```
 Enter-PSSession -Credential $creds -ComputerName x.x.x.x -Authentication Negotiate -Verbose
```
```
Enter-PSSession : Connecting to remote server x.x.x.x failed with the following error message : MI_RESULT_ACCESS_DENIED For more information, see the about_Remote_Troubleshooting Help topic.                                           
```

**Comment by SteveL-MSFT** [comment]
When remoting from Linux, you will likely need to explicitly use Basic auth.  Negotiate requires additional libraries and configuration.

**Comment by arnydo** [comment]
> When remoting from Linux, you will likely need to explicitly use Basic auth. Negotiate requires additional libraries and configuration.

Are the extra libraries required on the Windows or Linux side. Or both?

**Comment by SteveL-MSFT** [comment]
Windows already has the necessary libs built into the OS.  Here's an example where the docker image was updated with a lib to support NTLM on Linux: https://github.com/PowerShell/PowerShell-Docker/issues/124.  This [issue](https://github.com/PowerShell/PowerShell/issues/3329) seems to indicate Kerberos can work on Linux.

**Comment by arnydo** [comment]
> Windows already has the necessary libs built into the OS. Here's an example where the docker image was updated with a lib to support NTLM on Linux: [PowerShell/PowerShell-Docker#124](https://github.com/PowerShell/PowerShell-Docker/issues/124). This [issue](https://github.com/PowerShell/PowerShell/issues/3329) seems to indicate Kerberos can work on Linux.

Hmm. Checked to verify that the package is installed in the Docker image and still get the same issue as above.

```
PS /> apt show gss-ntlmssp
Package: gss-ntlmssp
Version: 0.7.0-4build3
Status: install ok installed
```

```
PS /> Enter-PSSession -Credential $cred -ComputerName xxxxx -Authentication Negotiate -Verbose
Enter-PSSession : Connecting to remote server xxxxx failed with the following error message : MI_RESULT_FAILED For more information, see the about_Remote_Troubleshooting Help topic
```

**Comment by RDIL** [comment]
Hey @arnydo,
All officially supported Docker images have `gss-ntlmssp` installed on them, except for `Alpine` and `openSUSE`.  So as long as you are using one of the others, you should be good.

**Comment by arnydo** [comment]
> Hey @arnydo,
> All officially supported Docker images have `gss-ntlmssp` installed on them, except for `Alpine` and `openSUSE`. So as long as you are using one of the others, you should be good.

I am using the latest Docker image for Powershell with no luck.

**Comment by danports** [comment]
@arnydo Did you check whether `/etc/services` was the issue? See https://github.com/PowerShell/PowerShell/issues/7342#issuecomment-477711182.

**Comment by arnydo** [comment]
> @arnydo Did you check whether `/etc/services` was the issue? See [#7342 (comment)](https://github.com/PowerShell/PowerShell/issues/7342#issuecomment-477711182).

Hey, what does /etc/services have to do with this particular issue? That isn't sarcastic...

**Comment by danports** [comment]
haha, yes, that's a fair question! There is an issue with the native OMI library PowerShell uses on Linux that only crops up with NTLM authentication when `/etc/services` is missing, as it is in some Docker images (like the official PowerShell images, IIRC). See https://github.com/microsoft/omi/issues/623.

**Comment by arnydo** [comment]
@danports I built a new image based on the latest Ubuntu 18.04 Powershell Dockerfile but with the addition of the ```RUN echo 'http 80/tcp www www-http' > /etc/services```.

Same result...

**Comment by danports** [comment]
Hmm, perhaps you're experiencing a different problem then. You might want to try [enabling OMI logging](https://github.com/microsoft/omi/blob/master/Unix/doc/diagnose-omi-problems.md#enabling-logging-in-omi) inside the container - reviewing the OMI logs is what eventually helped me to diagnose my issue.

**Comment by arnydo** [comment]
Doesn't look like omi is present at all in the Powershell images...is that even used in this case?

**Comment by jameskirsop** [comment]
@SteveL-MSFT, This is all well and good:

> When remoting from Linux, you will likely need to explicitly use Basic auth. Negotiate requires additional libraries and configuration.

Except, it's not supported:
```
> Enter-PSSession -Credential $creds -ComputerName <HOSTNAME> -Authentication Basic -Verbose
Enter-PSSession : Basic authentication is not supported over HTTP on Unix.
At line:1 char:1
....
```

You'd think that such a useful feature, with multiple bug reports, would have been implemented and/or documented properly a year and a half later after the issue was first reported...

**Comment by mgseelan** [comment]
Which version of the Power shell

**Comment by elandmancrs** [comment]
Just bumped against this too, using a rhel7.7 azure devops agent that needs to remote into a windows 2016 vm

**Comment by shneorc** [comment]
Same issue.
**Any solution or workaround?**
CentOS 7 --> Server 2016

```
PS /tmp> Enter-PSSession -ComputerName "HostName"
Enter-PSSession : MI_RESULT_ACCESS_DENIED
At line:1 char:1
+ Enter-PSSession -ComputerName "HostName"
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : InvalidArgument: (HostName:String) [Enter-PSSession], PSInvalidOperationException
+ FullyQualifiedErrorId : CreateRemoteRunspaceFailed
```

**Comment by MrHariSharma** [comment]
I am also facing same issue
CentOS 7 --> Windows Server 2019
PowerShell 6.0.0 Beta 6

PS /root> Enter-PSSession -ComputerName "TEST.mydomain.com"
Enter-PSSession : MI_RESULT_ACCESS_DENIED
At line:1 char:1
+ Enter-PSSession -ComputerName "TEST.mydomain.com"
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : InvalidArgument: (TEST.mydomain.com:String) [Enter-PSSession], PSInvalidOperationException
+ FullyQualifiedErrorId : CreateRemoteRunspaceFailed

**But if I try Invoke-Command it works**
**$dn='DomainName\Administrator'
$pwsd='GiveYourPassword'
$ip='192.168.20.119' 
$secpasswd = ConvertTo-SecureString $pwsd -AsPlainText -Force
$mycreds = New-Object System.Management.Automation.PSCredential -argumentlist $dn, $secpasswd**

**Now I have another interesting issue i.e. latency** 
Can someone please help?
If I use Powershell 6.2.3 RC build, Invoke-Command takes approx 30 sec. For more details #6531

**Comment by Nateb1121** [comment]
The reason this wasn't working for me was that I didn't have the NTLM authentication package ( gssntlmssp) for my distro (CentOS) installed. Rather than fiddling with installing the correct packages, I took a look at the official Microsoft PowerShell docker image. Currently the folks over there have been working on adding NTLM support into the docker container (https://github.com/PowerShell/PowerShell-Docker/issues/124) and have every platform but Photon which prevents it from being added to latest. 

**Here's a workaround** I used, YMMV. To run the PowerShell docker container on CentOS that should allow you to Enter-PSSession into Windows computers run the following 
`docker run -it mcr.microsoft.com/powershell:7.0.0-preview.6-centos-7`

I was able to successfully remote into things using that Docker container as a workaround. If you don't want to use the docker container then have a look at the Docker file in the PowerShell-Docker repo under `/release/preview/` for the corresponding architecture and see what commands it's running to get the container up. 

TL;DR Use the latest preview tagged Docker image for your architecture which will include the proper NTLM libraries. Alternatively, install gss-ntlmssp/gssntlmssp on your machine and it should solve most issues with Enter-PSSession.

**Comment by BitDesert** [comment]
Installing `gss-ntlmssp` as stated in #11374 solved this issue for me.
I'm using `-Authentication Negotiate`.

**Comment by iSazonov** [comment]
@SteveL-MSFT Please look previous comment.

**Comment by waljlopez2019** [comment]
> Installing `gss-ntlmssp` as stated in #11374 solved this issue for me.
> I'm using `-Authentication Negotiate`.

It's solution.....

**Comment by sliddjur** [comment]
I still have this issue, on Ubuntu 20.04
Powershell installed today with snap.
I'm trying to `Enter-PSSession -Authenthication Negotiate` using my $creds. 


```
~> snap list powershell
Name        Version  Rev  Tracking       Publisher              Notes
powershell  7.0.1    129  latest/stable  microsoft-powershell✓  classic
```

```
~> apt info gss-ntlmssp
Package: gss-ntlmssp
Version: 0.7.0-4build3
...
APT-Manual-Installed: yes
```

The error message is:
```
Enter-PSSession: Connecting to remote server 172.18.42.64 failed with the following error message : acquiring creds with username only failed Unspecified GSS failure.  Minor code may provide more information SPNEGO cannot find mechanisms to negotiate For more information, see the about_Remote_Troubleshooting Help topic.

```

**Comment by Himura2la** [comment]
Same here. Ubuntu 20.04, PowerShell 7.0.1 from Store:

```
PS /home/himura> Enter-PSSession -ComputerName ... -Credential ... -Authentication Negotiate

PowerShell credential request
Enter your credentials.
Password for user ...: ***************

Enter-PSSession: Connecting to remote server ... failed with the following error message : acquiring creds with username only failed Unspecified GSS failure.  Minor code may provide more information SPNEGO cannot find mechanisms to negotiate For more information, see the about_Remote_Troubleshooting Help topic.
```

**Comment by manivannanpk** [comment]
I am also facing the same issue with the latest version of PowerShell (7.0.1).

Trying to connect to Windows 2012 from CentOS Linux 7.

**Comment by manivannanpk** [comment]
As suggested by @BitDesert above, it works for me after installing gss-ntlmssp and with -Authentication Negotiate.

But there is a huge latency. Authentication itself takes around 20 seconds. Both the Linux and Windows machines are in the same subnet.

Any suggestion on how to reduce the latency?

**Comment by jlam55555** [comment]
Having the same issue here on Arch Linux with Powershell Core 7.0.0. Installing the [gss-ntlmssp AUR package][1] and using `-Authentication Negotiate` did not solve the problem. Still get:

```
PS /home/jon> Enter-PSSession -Credential $credentials -Authentication Negotiate -ComputerName 192.168.1.203
Enter-PSSession: Connecting to remote server 192.168.1.203 failed with the following error message : acquiring creds with username only failed Unspecified GSS failure.  Minor code may provide more information SPNEGO cannot find mechanisms to negotiate For more information, see the about_Remote_Troubleshooting Help topic.
```

[1]: https://aur.archlinux.org/packages/gss-ntlmssp/

**Comment by arizvisa** [comment]
Hey guys, op of PowerShell/PowerShell-Docker#124 here.

You can literally just use the docker://mcr.microsoft.com/powershell:centos-7 container w/ podman/docker/rkt/whatev to get it to work. Just run the container w/ interactivity and a tty and it kicks you into `pwsh`. If it doesn't work, then check that the server you're trying to connect to is configured properly because tokens aren't exchanged across both platforms. Not all platforms support gssapi/ntlm, and I don't think they're any tests..but because of @RDIL's work, it's super straightforward regardless.

```
Downloading sha256:d06345b12b6 [=============================]   106 MB / 106 MB 
Downloading sha256:524b0c1e57f [=============================] 75.9 MB / 75.9 MB 
PowerShell 7.0.3
Copyright (c) Microsoft Corporation. All rights reserved.

https://aka.ms/powershell
Type 'help' to get help.

PS /> enter-pssession -computername 10.7.17.218 -Credential $u -Authentication negotiate

PowerShell credential request
Enter your credentials.
User: root
Password for user root: ************************************************

[10.7.17.218]: PS C:\Users\root\Documents> exit
```

This is literally what containers are for, so you don't have to fight with crazy deps to get a simple task handled.

If you're trying to get kerberos to work with it (instead of ntlm), I don't think the containers will work (despite gssapi supporting it) because you'll need to get your tgt with `kinit` and then get the gssapi library to see it. You can probably do some clever mounting to get its library to see your tgt w/ the container though.

**Comment by MysticRyuujin** [comment]
I'd like to add that Linux Mint 20 has the same issue as Ubuntu 20.04.

I installed the package with snap and manually installed `gss-ntlmssp` via `apt` but I still get the error:
```
New-PSSession: [SERVER] Connecting to remote server SERVER failed with the following error message : acquiring creds with username only failed Unspecified GSS failure.  Minor code may provide more information SPNEGO cannot find mechanisms to negotiate For more information, see the about_Remote_Troubleshooting Help topic.
```

**Comment by se** [comment]
I found the problem. Just install `inetutils-ping` and it will solve. Probably it is depending `ping` to resolve OP or something.

```sh
apt-get install inetutils-ping
```

🖖

**Comment by sliddjur** [comment]
> I found the problem. Just install `inetutils-ping` and it will solve. Probably it is depending `ping` to resolve OP or something.
> 
> ```shell
> apt-get install inetutils-ping
> ```

On ubuntu I have `iputils-ping` - but it does not work with either one of them installed.

**Comment by janegilring** [comment]
I get the same error on Ubuntu 20.04 (running in WSL2), even though inetutils-ping and gss-ntlmssp is installed.

`Enter-PSSession:` This parameter set requires WSMan, and no supported WSMan client library was found. WSMan is either not installed or unavailable for this system.

**Comment by VGerris** [comment]
unfortunately it seems this kind of situations are terribly poorly documented by Microsoft.
I got the following to work:
 - Setup FreeIPA for DNS and Kerberos (on Linux)
 - register the window host in FreeIPA (add host - type machine name and IP address - save)
 - setup the proper ciphers for Kerberos (GPO/regedit, everything but DES)
 - Setup windows to authenticate with Kerberos, with something like:
```
 ksetup /setdomain IPA.YOURDOMAIN.COM
ksetup /addkdc IPA.YOURDOMAIN.COM idm.ipa.yourdomain.com
ksetup /addkpasswd IPA.YOURDOMAIN.COM idm.ipa.yourdomain.com
ksetup /setcomputerpassword SecretMachinePassword
ksetup /mapuser * *
```
  - Login to the FreeIPA host and generate the keytab:
  ipa-getkeytab -s ipa.yourdomain.com -p host/windows-hostname.ipa.yourdomain.com -e arcfour-hmac -k krb5.keytab.windows-hostname -P
  - now install a Linux host you will use to connect with Powershell - then register it with the FreeIPA Kerberos server and install powershell and the gssntlmssp package
  - on that same host you should be able to connect now:
  start powershell (pwsh / powershell in bash) then :
```
  $credential = get-credential user@IPA.YOURDOMAIN.COM
Enter-PSSession -ComputerName windows-hostname.ipa.yourdomain.com -Authentication Negotiate -Credential $credential
```
A few other notes : 
 - user needs to exist in FreeIPA, password is the password of the user in FreeIPA. The user also needs to exist as a local user on the Windows machine
 - while this worked from Linux, from Windows to Windows I only got it to work with Administrator accounts, then the format for the user is windows-hostname\Administrator
 - the Windows machines are showing the Kerberos Realm as a Workgroup, so they are not in a Domain like with an AD server
 - winrm service needs to run on host, setup and network has to be private (not public)
 ```
winrm quickconfig
Enable-PSRemoting
```
 
 Finally, the output on CentOS 7 to connect looks like :
 
```
 pwsh
PowerShell 7.1.3
Copyright (c) Microsoft Corporation.

https://aka.ms/powershell
Type 'help' to get help.

PS /home/user> $credential = get-credential user@IPA.YOURDOMAIN.COM

PowerShell credential request
Enter your credentials.
Password for user user@IPA.YOURDOMAIN.COM: ****************

PS /home/user> Enter-PSSession -ComputerName windows-hostname.ipa.yourdomain.com -Authentication Negotiate -Credential $credential
[windows-hostname.ipa.yourdomain.com]: PS C:\Users\user.WINDOWS-HOSTNAME\Documents>
```
All this was put together by pulling information from all over the net.
This does not contain all details but should give you enough info to get it to work.
I have not tried Basic authentication and I think it is not safe.
The better approach is to use SSH to manage Windows, because Microsoft seems to be unwilling to deliver proper cross platform authentication support (NTLM does not work and Kerberos not the same as in a domain setting).

There are tickets for that that have been open for months, I haven't checked those for a while, but feel free to try and get those done. Nobody wants to need a Windows server to manage other Windows servers ( or manage any windows server at all for that matter, but that's beyond the scope of the answer ;) ).

**Comment by VGerris** [comment]
> I still have this issue, on Ubuntu 20.04
> Powershell installed today with snap.
> I'm trying to `Enter-PSSession -Authenthication Negotiate` using my $creds.
> 
> ```
> ~> snap list powershell
> Name        Version  Rev  Tracking       Publisher              Notes
> powershell  7.0.1    129  latest/stable  microsoft-powershell✓  classic
> ```
> 
> ```
> ~> apt info gss-ntlmssp
> Package: gss-ntlmssp
> Version: 0.7.0-4build3
> ...
> APT-Manual-Installed: yes
> ```
> 
> The error message is:
> 
> ```
> Enter-PSSession: Connecting to remote server 172.18.42.64 failed with the following error message : acquiring creds with username only failed Unspecified GSS failure.  Minor code may provide more information SPNEGO cannot find mechanisms to negotiate For more information, see the about_Remote_Troubleshooting Help topic.
> ```

you are not posting your whole command line. Make sure to use hostname and not IP and set up credentials as shown in my post above. it seems you only supply a user. Just google that error and look at examples otherwise. Make sure to post full info if you have the same issue, thanks

**Comment by ponchofiesta** [comment]
```
PS > Enter-PSSession -ComputerName 192.168.10.85 -Credential $cred -Authentication Negotiate -UseSSL -SessionOption $sesopt
Enter-PSSession: Connecting to remote server 192.168.10.85 failed with the following error message : Authorization failed For more information, see the about_Remote_Troubleshooting Help topic.
```
The Windows eventlog shows Error code 0xC000035B. That might be a version problem. Powershell on Linux seems to use NTLMv1. Windows Server needs a newer version.

Client: CentOS 7, Powershell 7.1.3
Server: Windows Server 2019

**Comment by azurezhang** [comment]
In order to get the remote work, I have installed the following dependencies on ubuntu 20.4 (without Kerberos).
Install pwsh 7.1.3
Install OpenSSL
Install PSWSman : sudo pwsh -Command 'Install-Module -Name PSWSMan'
Install gss-ntlmssp : sudo apt-get install -y gss-ntlmssp

While I am trying to setup on RHEL with Kerberos, got the error of "acquiring creds with username only failed Unspecified GSS failure.  Minor code may provide more information SPNEGO cannot find mechanisms to negotiate" from Invoke-Command, and "Authorization failed Unspecified GSS failure.  Minor code may provide more information Server not found in Kerberos database For more information" from New-PSSession and Enter-PSSession command

This might be a Kerberos configure issue or still missing some other dependencies required on RHEL.

**Comment by VGerris** [comment]
you need to have your Linux machine authenticate with Kerberos first. Depending if you use AD or Linux, you need to make sure that it works. As I wrote, it works with both the Windows and Linux host using a Linux Kerberos server.
So login on the machine with Kerberos and  type klist, you should see a valid ticket there.
As far as I know, it will not work without that, correct me if I'm wrong.

**Comment by azurezhang** [comment]
@VGerris Thank you for your comment, I am able to create/enter PSSession from Linux to windows server run after turn off FIPS on Linux. Tracing down to md5 hash error and find that gss-ntlmssp is not compatible with FIPS.
Now need found a FIPS compatible gss-ntlmssp package.

**Comment by VOVELEE** [comment]
I am experiencing similar issues with my setup - I simply cannot make Ubuntu 18.04 or 20.04 to connect to Windows Server 2019 using NTLM. Does anyone have any suggestions?

Looks like Ubuntu and Windows Server 2019 cannot negotiate the correct NTLM.
The same command opens a valid PSSession when it is executed on Windows 10 Client (standalone client, not joined to doimain).

Setup:

- Enabled WinRM on Windows Server 2019. Windows Server is joined to an Azure Active Directory Domain Services domain
- Installed Powershell 7.1.3 on Ubunbtu 18.04 following official Microsoft article
- Installed gss-ntlmssp on Ubuntu 18.04 to enable NTML authentication as per https://github.com/PowerShell/PowerShell-Docker/issues/124
- Install PSWSMAN 2.2.0 (I tested it without it - the same issue occur)

Commands executed on Ubuntu server:

``` powershell
$PSSessionParameters = @{
  Authentication    = 'Negotiate'
  Credential        = [pscredential]::new('user@domaintest.onmicrosoft.com', ('Obfuscated' | ConvertTo-SecureString -AsPlainText))
  UseSSL            = $true
  Port              = 5986
  ConfigurationName = 'PowerShell.7'
  SessionOption     = New-PSSessionOption -SkipCACheck -SkipCNCheck
}
New-PSSession -ComputerName '10.0.104.201' @PSSessionParameters
```

Error received on Ubuntu: 

```
New-PSSession: [10.0.104.201] Connecting to remote server 10.0.104.201 failed with the following error message : Authorization failed For more information, see the about_Remote_Troubleshooting Help topic.
```


Error in Windows Server Security log:
```
Event 4625
Account For Which Logon Failed:
	Security ID:		NULL SID
	Account Name:		user@domaintest.onmicrosoft.com
	Account Domain:		

Failure Information:
	Failure Reason:		An Error occured during Logon.
	Status:			0xC000035B
	Sub Status:		0x0
```

NTLM audit log on Windows Server:

```
NTLM server blocked audit: Audit Incoming NTLM Traffic that would be blocked
Calling process PID: 3136
Calling process name: C:\Windows\System32\svchost.exe
Calling process LUID: 0x3E4
Calling process user identity: vm-gs-alt001$
Calling process domain identity: DOMAINTEST
Mechanism OID: 1.3.6.1.4.1.311.2.2.10
Audit NTLM authentication requests to this server that would be blocked if the security policy Network Security: Restrict NTLM: Incoming NTLM Traffic is set to Deny all accounts or Deny all domain accounts.
If you want this server to allow NTLM authentication, set the security policy Network Security: Restrict NTLM: Incoming NTLM Traffic to Allow all.
```

**Comment by VOVELEE** [comment]
Please take a look into this threat - https://github.com/jborean93/omi/issues/29
Looks like the limitation comes from the libraries which ship with different version of Windows. The workaround is to set the CbtHardeningLevel to None so Windows doesn't try to validate the value at all.

**Comment by netcore-jroger** [comment]
if you using `CentOS`, please install gssntlmssp:
```bash
yum install gssntlmssp
```
if using `Ubuntu`, please install gss-ntlmssp:
```bash
sudo apt install gss-ntlmssp
```

**Comment by plao** [comment]
> New-PSSession -ComputerName remotehost -Credential $credential -Authentication Negotiate
> 
> Use Negotiate for authentication. I don't recommend using basic as that would send your password in plain text over the network and you would also need to manually allow the remote host to accept basic authentication as it is not allowed by default. Specifying Negotiate results in using NTLM authentication which can be verified by checking $PSSenderInfo.UserInfo.Identity.AuthenticationType

Thanks!

**Comment by hunter86bg** [comment]
Obviously something is wrong with the PowerShell for Linux.
A very interesting [blog](https://www.bloggingforlogging.com/2020/08/21/wacky-wsman-on-linux/) that can shed the light , especially the topic where the PRs were rejected.

For my Ubuntu18.04 connecting to Win10 Pro over winrm (http port, not joined in AD) , the following worked:

```
sudo apt install  gss-ntlmssp powershell
pwsh -Command 'Install-Module -Name PSWSMan'
sudo pwsh -Command 'Install-WSMan'
```

Validation:
```
$cred=Get-Credential 
Enter-PSSession -ComputerName vmhostname.vmdomain -Credential $cred -Authentication Negotiate
```

File copy :
```
$pw = convertto-securestring -AsPlainText -Force -String PASS
$cred = new-object -typename System.Management.Automation.PSCredential -argumentlist "USER",$pw
$session = New-PSSession -ComputerName win10pro.localdomain -Credential $cred  -Authentication Negotiate
Copy-Item -Path 'C:\Users\USER\Desktop\somefile.txt' -Destination /tmp/ -FromSession $session
```

**Comment by VarunRajanna** [comment]
The below steps worked for me on CentOS 7*
 
Install PowerShell for linux 
DOCS: https://docs.microsoft.com/en-us/powershell/scripting/install/installing-powershell-on-linux?view=powershell-7.2
 
Register the Microsoft RedHat repository
curl https://packages.microsoft.com/config/rhel/8/prod.repo | sudo tee /etc/yum.repos.d/microsoft.repo
    
Install PowerShell

``` sudo yum install -y powershell ```
 
Confirm epel-release is installed

``` yum install -y epel-release ```
 
Update all packages
 
``` yum update -y ```
 
Install the below packages

``` yum install -y gssntlmssp less ncurses openssh-clients ```
 
Note: If gssntlmssp fails with the error Public key for gssntlmssp-0.7.0-1.el7.x86_64.rpm is not installed use the below command

``` yum install -y gssntlmssp --nogpgcheck ```
 
Install the PSWSMan on powershell

``` pwsh -Command "Install-Module -Name PSWSMan -Force ```

**Comment by arizvisa** [comment]
> Obviously something is wrong with the PowerShell for Linux. A very interesting [blog](https://www.bloggingforlogging.com/2020/08/21/wacky-wsman-on-linux/) that can shed the light , especially the topic where the PRs were rejected.

For the record, the blog is definitely titled as being related to Linux..but if you look at the PRs microsoft/omi#669, microsoft/omi#670, and the blog, those are all related to building the library on MacOS (which is unsupported).

The troubleshooting of GSSAPI and everything else from the blog is definitely on-topic, though, and it's super cool that the author is actively maintaining OMI for all of the platforms (including MacOS)...because honestly, it needs it.

**Comment by hunter86bg** [comment]
> > Obviously something is wrong with the PowerShell for Linux. A very interesting [blog](https://www.bloggingforlogging.com/2020/08/21/wacky-wsman-on-linux/) that can shed the light , especially the topic where the PRs were rejected.
> 
> For the record, the blog is definitely titled as being related to Linux..but if you look at the PRs [microsoft/omi#669](https://github.com/microsoft/omi/pull/669), [microsoft/omi#670](https://github.com/microsoft/omi/pull/670), and the blog, those are all related to building the library on MacOS (which is unsupported).
> 
> The troubleshooting of GSSAPI and everything else from the blog is definitely on-topic, though, and it's super cool that the author is actively maintaining OMI for all of the platforms (including MacOS)...because honestly, it needs it.

And he is also an Ansible developer (from Red Hat) focusing on Windows.

**Comment by hunter86bg** [comment]
It's really interesting , as even 'install Install-WSMan' doesn't work when I want to copy from Linux to Windows 10 Pro. 'Copy-Item -ToSession ' just stucks.
Using pywinrm (python) is a kind of solution , but it doesn't support file transfer, so you have to base64 encode on one side and base64 decode on the other side.

**Comment by arizvisa** [comment]
@hunter86bg not to police this thread, but that might be off-topic since we're in an issue tracker for keeping track of bugs and the thread could get closed or locked if it wanders too far from the original issue. this specific issue is with regards to entering/creating a PSsession and not necessarily performance issues encountered therein.

please create another issue describing your problem so that maybe the devers could look into it.

**Comment by lefort1** [comment]
I spent days trying to figure out why the code below wasn't working on **Amazon Linux 2 (CentOS 7)**:
```
$serviceUserName = '<username>@my.company.domain'
$secStringPassword = '<password>'
​$secStringPassword = ConvertTo-SecureString $servicePassword -AsPlainText -Force
$credObject = New-Object System.Management.Automation.PSCredential ($serviceUserName, $secStringPassword)

$ExSession = New-PSSession –ConfigurationName Microsoft.Exchange –ConnectionUri ‘http://<hostname>.<domain>/PowerShell/?SerializationLevel=Full’ -Credential $credObject -Authentication Kerberos
```

Error message:
```
Enter-PSSession : Connecting to remote server hostname failed with the following error message : Kerberos verify cred with password failed No credentials were supplied, or the credentials were unavailable or inaccessible For more information, see the about_Remote_Troubleshooting Help topic.
```

Turns out the domain name in the username simply needed to be capitalized like so: `<username>@MY.COMPANY.DOMAIN`

Sorry if this is obvious, but hoping it helps someone else like me.

**Comment by mtbiker-s** [comment]
> > Installing `gss-ntlmssp` as stated in #11374 solved this issue for me.
> > I'm using `-Authentication Negotiate`.
> 
> It's solution.....

This worked for me on Centos 7 with Powershell 7.1.3

**Comment by ponchofiesta** [comment]
> 



> > > Installing `gss-ntlmssp` as stated in #11374 solved this issue for me.
> > > I'm using `-Authentication Negotiate`.
> > 
> > 
> > It's solution.....
> 
> This worked for me on Centos 7 with Powershell 7.1.3

It's a Debian package and not available in CentOS 7.

**Comment by ysakkk** [comment]
My Case.

* PowerShell command from python pywinrm

  * Environment
    * Hyper-V
      - Windows Server 2016 Datacenter Evaluation
        IP:  172.20.137.90
        AD: test.local
  
    * WSL  
      - Ubuntu 22.04.2
        IP: 172.17.173.189

  * Allow Access WSL to Hyper-V
`    PowerShell> Get-NetIPInterface -InterfaceAlias 'vEthernet (Default Switch)', 'vEthernet (WSL (Hyper-V firewall))' | Set-NetIPInterface -Forwarding Enabled`

* winrm_ps.py
```
  import winrm
  endpoint = "172.20.137.90:5985"
  username = 'test\\administrator'
  password = 'XXXXXXX#######'
  s = winrm.Session(endpoint, auth=(username, password), transport='ntlm')
  r = s.run_ps("hostname")
  print(r.std_out.decode('utf-8'))
  print(r.std_err.decode('utf-8'))
```

  $ python3 winrm_ps.py
  WIN-RPIEQXFIST
